engine.lisp

```diff
@@ -87,6 +87,90 @@ for the selector to the matcher."))
                      (incf end)
                   finally (test) (return NIL))))))))
 
+;; https://drafts.csswg.org/cssom/#serialize-an-identifier
+(defun css-escape (string)
+  "Escape all the invalid CSS characters to their safe counterparts."
+  (declare (optimize speed)
+           (type string string))
+  (let ((buffer (make-array 0 :adjustable t :fill-pointer 0)))
```

Comment:
Use a string buffer (aka `make-string-output-stream`), or even better, `with-output-to-string`.